### PR TITLE
gate rustc_on_unimplemented under rustc_attrs

### LINKED ISSUE #29628: Tracking issue for `on_unimplemented` feature
The `on_unimplemented` feature provides the `#[rustc_on_unimplemented]` attribute, which allows trait definitions to add specialized notes to error messages when an implementation was expected but not found.

**WARNING:** This attribute, as all other `rustc_` attributes, is an internal implementation detail of `rustc` and is not meant to become stable.

### PR DESCRIPTION
**Pull request by Centril**
Move `rustc_on_implemented` from the `on_implemented` gate to `rustc_attrs` as it is internal.

Closes #29628

r? @varkor

### CODE REVIEW

**Comment by varkor** [comment]
We should remove the feature gate too: otherwise it's possible to enable the feature without any effect.

**Comment by Centril** [comment]
@varkor I don't know what you mean by "without any effect". The only technical change is the omission of the tracking issue. Would you prefer to move it to rustc_attrs?

**Comment by varkor** [comment]
I mean that you can currently write:
```rust
#![feature(on_unimplemented)]
```
And there won't be any error, which there ought to be after this change.

**Comment by Centril** [comment]
@varkor I didn't intend for that to be an error, only to remove any mention of the tracking issue. The attribute has to be gated by something, either the existing feature gate name or rustc_attrs. Do we want to gate all internal features with rustc_attrs? The current practice seems to be all over the place.

**Comment by varkor** [comment]
What other `rustc_*` attributes use their own feature gate rather than `rustc_attrs`?

**Comment by Centril** [comment]
Seems like it's a small minority:

- rustc_const_unstable
- rustc_diagnostic_item
- rustc_paren_sugar

(`rustc_` attributes are not the only internal attributes)

**Comment by varkor** [comment]
Do you know if there's any particular reason for those? I'll investigate later if not. I would have thought that all `rustc_*` attributes could come under the same feature gate.

**Comment by Centril** [comment]
> Do you know if there's any particular reason for those?

No idea but my guess is that it would be happenstance.

**Comment by varkor** [comment]
As far as I can tell, `rustc_paren_sugar` and `rustc_const_unstable` should be gated on `rustc_attrs`.
- `rustc_diagnostic_item` already is gated on `rustc_attrs`.

So I think we should gate `rustc_on_unimplemented`  on `rustc_attrs` too (and change the others either here, or in a follow up).

**Comment by Centril** [comment]
@varkor Moved `rustc_on_unimplemented` for now. :)

**Comment by Centril** [comment]
@varkor I think it's good to go now.

**Comment by JohnCSimon** [comment]
Ping from triage.
@varkor    Can you please review this PR?
cc: @Centril 

Thank you!

File: src/libsyntax/feature_gate/builtin_attrs.rs [hunk 1]
```diff
@@ -167,7 +167,7 @@ macro_rules! experimental {
 }
 
 const IMPL_DETAIL: &str = "internal implementation detail";
-const INTERAL_UNSTABLE: &str = "this is an internal attribute that will never be stable";
+const INTERNAL_UNSTBLE: &str = "this is an internal attribute that will never be stable";
```
Comment (varkor) [inline src/libsyntax/feature_gate/builtin_attrs.rs]: Why did this name change?

Comment (Centril) [reply src/libsyntax/feature_gate/builtin_attrs.rs]: A typo fix

Comment (varkor) [reply src/libsyntax/feature_gate/builtin_attrs.rs]: It looks like it changed from one typo to another. Shouldn't it be `INTERNAL_UNSTABLE`?

Comment (Centril) [reply src/libsyntax/feature_gate/builtin_attrs.rs]: Argh.. Didn't notice. Will fix. Can you review the rest of the PR meanwhile?

**Comment by varkor** [comment]
r=me with typo fixed.

**Comment by Centril** [comment]
Typo fixed; @bors r=varkor rollup

**Pull request merged**